### PR TITLE
[WIP] Introduce GlutenOutputWriterFactory to unify DSv1 writer

### PR DESCRIPTION
**Pull request by Yohahaha**
## What changes were proposed in this pull request?

Since PR #1926 propose a great idea to support offloaded writer to Gluten, this PR will further improve it.

### Current limitation

1. Only support datasource v1 interface.
2. Only support parquet format.
3. Gluten fat jar need pre loaded in Spark, cause we rewrite some class belongs to Spark.

## How was this patch tested?

Since current UT framework could not pre load rewritten classes, it can only be tested in Spark application for now.

### CODE REVIEW

**Comment by Yohahaha** [comment]
I get some great ideas from your issue and PR, and ETL scenario support is important to us, hope you don't mind me submit this PR. We could improve writer support if you like, thanks! @binmahone

**Comment by zhouyuan** [comment]
@JkSelf

File: shims/spark33/src/main/scala/org/apache/spark/sql/hive/execution/HiveFileFormat.scala [hunk 1]
```diff
@@ -0,0 +1,200 @@
...
+ * This class is copied from Spark 3.3.1 and modified for Gluten. Gluten should make sure this class
+ * is loaded before the original class.
+ */
+
+/**
+ * `FileFormat` for writing Hive tables.
+ *
+ * TODO: implement the read logic.
+ */
+class HiveFileFormat(fileSinkConf: FileSinkDesc)
```
Comment (JkSelf) [inline shims/spark33/src/main/scala/org/apache/spark/sql/hive/execution/HiveFileFormat.scala:57]: Can put these classes into common folder?

Comment (Yohahaha) [reply shims/spark33/src/main/scala/org/apache/spark/sql/hive/execution/HiveFileFormat.scala:57]: HiveFileFormat of v3.3.1 has a new method `supportFieldName` compared with v3.2.2

**Pull request closed without merging**